### ISSUE TITLE
concrete_solve: How to detect if solution was succesful? Or how to get retcode?

### ISSUE THREAD

**Issue by scheidan** [SciML/SciMLSensitivity.jl#234]
During optimization we may hit an unstable region in the parameter space. The classic workaround is to return `-Inf` as nicely documented [here](https://docs.sciml.ai/dev/analysis/parameter_estimation/#Note-About-Loss-Functions-1). Unfortunately `concrete_solve` doesn't give a `retcode`.
So I'm wondering:
- How can we "catch" potential instability warnings in the loss functions?
- Would retuning `-Inf` be a good idea at all? It seems to me that this would break gradient descent.

Thanks a lot!

**Comment by ChrisRackauckas** [comment]
I think we need to find a nice safe way for concrete_solve to return the original solution object, but in a safe way so that if someone wants to interpolate it properly errors (since those gradients won't propagate). Basically, instead of going to DiffEqArray, it should scrub sol.k and change the interpolant to linear (since that's okay for derivatives) and return this linear interpolation only solution. IT's not too difficult to do, but it would take some work and we'd want to test it well to not leave users with tricky edge cases that give 0 gradients.

**Comment by ChrisRackauckas** [comment]
https://github.com/SciML/DifferentialEquations.jl/issues/610 is towards a real solution. I'll close this and we can track there.

**Comment by ArnoStrouwen** [comment]
Did getting rid of concrete solve fix this issue entirely?
```jl
using DiffEqSensitivity
using OrdinaryDiffEq
using ReverseDiff
using Zygote

function fiip(du,u,p,t)
    du[1] = dx = p[1]*u[1] - p[2]*u[1]*u[2]
    du[2] = dy = -p[3]*u[2] + p[4]*u[1]*u[2]
end

function sum_of_solution(p)
    _prob = remake(prob,p=p)
    sol = solve(_prob,Tsit5(),saveat=0.1,sensealg=QuadratureAdjoint())
    if sol.retcode != :Success
        return p[1]
    end
    sum(sol)
end
p = [1.5,1.0,3.0,1.0]; u0 = [NaN;1.0]
prob = ODEProblem(fiip,u0,(0.0,10.0),p)
dp = ReverseDiff.gradient(sum_of_solution,p) # does not work
dp = Zygote.gradient(sum_of_solution,p) # works

```

**Comment by ChrisRackauckas** [comment]
Yes it did.

**Comment by ChrisRackauckas** [comment]
Oh, that last piece is an upstream issue: ReverseDiff.jl cannot allow derivative rules to return general structs, so we have to return an array. That's a ReverseDiff.jl issue though.

**Comment by ChrisRackauckas** [comment]
concrete_solve is gone, so the retcode exists.
